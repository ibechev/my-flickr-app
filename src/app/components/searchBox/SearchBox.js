import React from "react";
import PropTypes from "prop-types";

import Tag from "../tag/Tag";
import Button from "../button/Button";

export const SearchBox = ({
  inputFocus,
  inputValue,
  error,
  tags,
  refs,
  ...handle
}) => {
  const focusClass = inputFocus ? "focus" : "";
  const errorClass = error ? "error" : "";
  const placeholder = "... Enter tags here. Separate with space or comma";

  return (
    <section
      id="search-box"
      className={`search-box ${errorClass} ${focusClass}`}
      onClick={handle.searchBoxFocus}
      ref={refs.searchBox}
      data-test="search-box"
    >
      {error && (
        <span className="error" data-test="input-error">
          {error}
        </span>
      )}

      <form action="" onSubmit={e => handle.addTag(e, inputValue.trim())}>
        <section className="tags">
          <ul className="tags-wrapper">
            {tags.map((tag, i) => (
              <Tag
                key={tag}
                index={i}
                value={tag}
                removeTag={handle.removeTag}
              />
            ))}

            <li className="tags-input-wrapper">
              <input
                id="tags-input"
                type="text"
                className="tags-input"
                size="4"
                autoComplete="off"
                onKeyDown={handle.keyDown}
                onChange={handle.change}
                autoFocus
                ref={refs.inputEl}
                value={inputValue}
              />
            </li>
          </ul>

          {!tags.length && !inputValue.length && (
            <span className="placeholder" data-test="tags-placeholder">
              {placeholder}
            </span>
          )}
        </section>

        <section className="controls">
          <Button
            className="clear"
            onClick={handle.removeAllTags}
            title="Clear all tags"
            type="button"
            data-test="btn-clear-tags"
          >
            <i className="far fa-trash-alt" />
            Clear
          </Button>
        </section>
      </form>
    </section>
  );
};

SearchBox.propTypes = {
  tags: PropTypes.arrayOf(PropTypes.string),
  inputValue: PropTypes.string,
  inputFocus: PropTypes.bool.isRequired,
  error: PropTypes.string,
  refs: PropTypes.shape({
    searchBox: PropTypes.shape({
      current: PropTypes.node.isRequired.isRequired
    }).isRequired,
    inputEl: PropTypes.shape({
      current: PropTypes.node.isRequired.isRequired
    }).isRequired
  }),
  handle: PropTypes.shape({
    searchBoxFocus: PropTypes.func.isrequired,
    keyDown: PropTypes.func.isrequired,
    addTag: PropTypes.func.isrequired,
    change: PropTypes.func.isrequired,
    search: PropTypes.func.isrequired,
    removeTag: PropTypes.func.isrequired,
    removeAllTags: PropTypes.func.isrequired
  })
};

export default SearchBox;
