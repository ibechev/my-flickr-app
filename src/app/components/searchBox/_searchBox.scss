.search-box {
  position: relative;
  margin: 4rem auto 5rem;
  max-width: 600px;
  padding: 0.7rem 1rem;
  @include border-radius(5px);
  background-color: $primaryVeryLight;
  border: 1px solid $primary;
  @include box-shadow($shadow);
  transition: border 0.1s ease-in-out;

  &.focus {
    border: 1px solid darken($primary, 20%);
  }

  &.error {
    border: 1px solid $warning;
  }

  .tags {
    position: relative;
    margin-bottom: 0.7rem;

    &:hover {
      cursor: text;
    }

    .tags-wrapper {
      @include flex;
      @include align-items(center);
      @include flex-wrap(wrap);
      padding-top: 0.35rem;

      .tags-input-wrapper {
        @include flex;
        @include align-items(center);
        margin-bottom: 0.35rem;

        .tags-input {
          background-color: transparent;
          line-height: 1.4;
          font-size: 1.8rem;
        }
      }
    }
    .placeholder {
      position: absolute;
      top: 50%;
      @include transform(translateY(-50%));
      line-height: 1.6;
      color: $primaryDark;
    }
  }

  .controls {
    @include flex;
    @include justify-content(flex-end);
    padding: 1.4rem 0 0.7rem;
    border-top: 1px solid $primary;

    button {
      padding: 0 0.7rem;
      background-color: transparent;
      font-size: 1.8rem;
      i {
        margin-right: 0.5rem;
      }

      &:hover {
        cursor: pointer;
      }

      &.clear {
        color: $warning;

        &:hover {
          color: $warningHover;
        }
      }

      &.search {
        color: $success;

        &:hover {
          color: $successHover;
        }
      }
    }
  }

  .error {
    position: absolute;
    right: 0;
    bottom: 105%;
    font-size: 1.2rem;
    color: $warning;
  }

  @media screen and (max-width: 640px) {
    margin: 3rem 1.5rem 5rem;
  }
}
