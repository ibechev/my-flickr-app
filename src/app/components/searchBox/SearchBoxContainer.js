import React, { useState, useEffect, useRef } from "react";
import PropTypes from "prop-types";
import { connect } from "redhooks";
import queryString from "query-string";
import { validateInput, validateSearch } from "../../utils/validation";
import {
  addTag,
  addTags,
  removeTag,
  removeAllTags,
  clearImages,
  fetchImagesAction
} from "../../actions/index";
import { encodeURITags } from "../../utils";

import {
  START_SEARCH,
  SEARCH_COMPLETE,
  FETCH_IMAGES_ERROR
} from "../../actions/types";

import SearchBox from "./SearchBox";

const SearchBoxContainer = ({
  tags,
  addTag,
  addTags,
  removeTag,
  removeAllTags,
  clearImages,
  fetchImagesAction,
  location,
  history
}) => {
  const [inputValue, setInputValue] = useState("");
  const [inputFocus, setInputFocus] = useState(true);
  const [error, setError] = useState("");
  const refs = {
    searchBox: useRef(null),
    inputEl: useRef(null)
  };

  useEffect(() => {
    const paramName = "tags";
    const query = queryString.parse(location.search);

    if (query === undefined || query[paramName] === undefined) return;

    query[paramName] = query[paramName].trim();

    if (query[paramName].length > 0) {
      addTags(query[paramName].split(" "));
    }
  }, []);

  useEffect(() => {
    window.addEventListener("click", clickListener, false);

    return function clearEventListener() {
      window.removeEventListener("click", clickListener, false);
    };
  });

  /**
   * Equivalent to componentDidUpdate
   */
  const isMounted = useRef(false);
  useEffect(() => {
    // Prevent the execution of the functions on the first render
    if (!isMounted.current) {
      isMounted.current = true;
      return;
    }

    if (tags.length !== 0) {
      handleSearch(tags, error, setError);
      history.replace(`/?tags=${encodeURITags(tags).join("%20")}`);
    } else {
      clearImages();
      history.replace("/");
    }
  }, [tags]);

  const handleSearch = (tags, error, setError) => {
    const tmpError = validateSearch(tags);

    if (tmpError) {
      setError(error);
      return;
    }

    // Clear old errors, if there are no new ones
    !tmpError.length && error.length && setError("");

    fetchImagesAction({
      type: START_SEARCH,
      meta: {
        success: SEARCH_COMPLETE,
        error: FETCH_IMAGES_ERROR
      }
    });

    clearImages();
  };

  const handleKeyDown = e => {
    const { keyCode } = e;

    /**
     * If Space or Comma are pressed, add the input as a new tag
     */
    switch (keyCode) {
      case 32: // Space
      case 188: // Comma
        handleAddTag(e, inputValue.trim());
        break;

      case 8: // Backspace - remove tag
        !inputValue.trim().length && removeTag();
        break;

      default:
        /**
         * When typing and deleting letters increase
         * or decresce the size of the input element
         */
        adjustInputWidth(e.target, inputValue);
    }
  };

  const handleAddTag = (e, inputValue) => {
    e.preventDefault();
    const error = validateInput(inputValue, tags);

    if (error.length) {
      setError(error);
      return;
    }

    !error.length && error.length && setError("");

    if (inputValue.length) {
      addTag(inputValue);
      setInputValue("");
    }
  };

  const handleChange = e => {
    setInputValue(e.target.value.trim());
    !inputValue.length && !!error.length && setError("");

    // Change the size of the input element on text paste
    adjustInputWidth(e.target, inputValue);

    // Remove any existing error
    !!error.length && setError("");
  };

  const adjustInputWidth = (input, value) =>
    (input.size = value.length > 1 ? value.length + 2 : 4);

  const hanldeSearchBoxFocus = () => {
    if (!inputFocus) {
      setInputFocus(true);
    }
    refs.inputEl.current.focus();
  };

  const clickListener = e => {
    /**
     * Remove the focus from the input, if the click is outside the search box
     */
    if (!inputFocus) return;

    if (
      e.target !== refs.searchBox.current &&
      !refs.searchBox.current.contains(e.target)
    )
      setInputFocus(false);
  };

  return (
    <SearchBox
      inputValue={inputValue}
      inputFocus={inputFocus}
      error={error}
      tags={tags}
      refs={refs}
      searchBoxFocus={hanldeSearchBoxFocus}
      keyDown={handleKeyDown}
      addTag={handleAddTag}
      change={handleChange}
      search={handleSearch}
      removeTag={removeTag}
      removeAllTags={removeAllTags}
    />
  );
};

const mapStateToProps = ({ tags }) => ({
  tags
});

SearchBoxContainer.propTypes = {
  tags: PropTypes.arrayOf(PropTypes.string),
  fetchImagesAction: PropTypes.func.isRequired,
  addTag: PropTypes.func.isRequired,
  addTags: PropTypes.func.isRequired,
  removeTag: PropTypes.func.isRequired,
  removeAllTags: PropTypes.func.isRequired,
  clearImages: PropTypes.func.isRequired,
  location: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired
};

export default connect(
  mapStateToProps,

  { addTag, addTags, removeTag, removeAllTags, clearImages, fetchImagesAction }
)(SearchBoxContainer);
