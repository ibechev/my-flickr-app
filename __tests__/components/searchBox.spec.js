import React, { createRef } from "react";
import renderer from "react-test-renderer";
import { SearchBox } from "../../src/app/components/searchBox/SearchBox";
import { propTypeErrors } from "../testUtils";
import { testTags } from "../testValues";
import { mount } from "enzyme";

describe("Expect <SearchBox />", () => {
  const defaultProps = {
    inputValue: "",
    inputFocus: true,
    error: "",
    tags: [],
    refs: { searchBox: createRef(), inputEl: createRef() },
    searchBoxFocus: () => {},
    keyDown: () => {},
    addTag: () => {},
    change: () => {},
    search: () => {},
    removeTag: () => {},
    removeAllTags: () => {}
  };

  const setup = (props = {}) => {
    const wrapper = mount(<SearchBox {...defaultProps} {...props} />);
    return wrapper;
  };

  it("to render correctly", () => {
    const wrapper = renderer.create(<SearchBox {...defaultProps} />).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it("to not throw error with correct props", () => {
    propTypeErrors(<SearchBox {...defaultProps} />, defaultProps);
  });

  it("to not display any <Tag /> conponent, if 'state.tags' is empty", () => {
    const wrapper = setup(defaultProps);
    expect(wrapper.find("Tag")).toHaveLength(0);
  });

  it("to display number of <Tag /> components, equal of the number of tags in 'props.tags'", () => {
    const wrapper = setup({ tags: [...testTags] });
    expect(wrapper.find("Tag")).toHaveLength(testTags.length);
  });

  it("to display <span data-test='tags-placeholder' /> only, if 'props.tags' is empty and/or 'props.inputValue.length' eauals 0", () => {
    const testPropsOne = {
      tags: [...testTags]
    };
    const testPropsTwo = {
      inputValue: "abc"
    };
    const testPropsThree = {
      tags: [...testTags],
      inputValue: "abc"
    };
    const wrapper = setup();
    const wrapperOne = setup(testPropsOne);
    const wrapperTwo = setup(testPropsTwo);
    const wrapperThree = setup(testPropsThree);

    const spanName = "span[data-test='tags-placeholder']";

    expect(wrapper.find(spanName)).toHaveLength(1);
    expect(wrapperOne.find(spanName)).toHaveLength(0);
    expect(wrapperTwo.find(spanName)).toHaveLength(0);
    expect(wrapperThree.find(spanName)).toHaveLength(0);
  });

  it("to add class 'focus' to <section data-test='search-box' />, if 'props.inFocus' equals 'true'", () => {
    const wrapperWithFocus = setup();
    const wrapperWithoutFocus = setup({ inputFocus: false });

    const sectionName = "section[data-test='search-box']";
    const focusClass = "focus";

    expect(wrapperWithFocus.find(sectionName).hasClass(focusClass)).toEqual(
      true
    );
    expect(wrapperWithoutFocus.find(sectionName).hasClass(focusClass)).toEqual(
      false
    );
  });

  it("to display <span data-test='input-error' /> and the error text inside, only if 'props.error' is not an empty string", () => {
    const testProps = {
      error: "Some error message"
    };

    const wrapper = setup();
    const spanName = "span[data-test='input-error']";
    expect(wrapper.find(spanName)).toHaveLength(0);

    const wrapperWithError = setup(testProps);
    expect(wrapperWithError.find(spanName)).toHaveLength(1);
    expect(wrapperWithError.find(spanName).text()).toEqual(testProps.error);
  });

  it("to call 'props.removeAllTags', when <button dats-test='clear-tags' is clicked", () => {
    const removeAllTags = jest.fn();
    const wrapper = setup({ removeAllTags });

    wrapper.find("button[data-test='btn-clear-tags']").simulate("click");
    expect(removeAllTags).toHaveBeenCalledTimes(1);
  });

  it("to call 'props.addTag' only when 'props.inputValue' is not an empty string AND 'props.error' IS an empty string", () => {});
});
