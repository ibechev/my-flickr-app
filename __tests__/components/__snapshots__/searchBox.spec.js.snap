// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <SearchBox /> to render correctly 1`] = `
<section
  className="search-box  focus"
  data-test="search-box"
  id="search-box"
  onClick={[Function]}
>
  
  <form
    action=""
    onSubmit={[Function]}
  >
    <section
      className="tags"
    >
      <ul
        className="tags-wrapper"
      >
        <li
          className="tags-input-wrapper"
        >
          <input
            autoComplete="off"
            autoFocus={true}
            className="tags-input"
            id="tags-input"
            onChange={[Function]}
            onKeyDown={[Function]}
            size="4"
            type="text"
            value=""
          />
        </li>
      </ul>
      <span
        className="placeholder"
        data-test="tags-placeholder"
      >
        ... Enter tags here. Separate with space or comma
      </span>
    </section>
    <section
      className="controls"
    >
      <button
        className="btn clear"
        data-test="btn-clear-tags"
        onClick={[Function]}
        title="Clear all tags"
        type="button"
      >
        <i
          className="far fa-trash-alt"
        />
        Clear
      </button>
    </section>
  </form>
</section>
`;
